### ISSUE TITLE
[VL] use multiple threads in the same executor

### ISSUE THREAD

**Issue by FelixYBW** [apache/incubator-gluten#7810]
### Description

Velox is initially designed to use multiple threads in each context. To fit spark's working model, we limit 1 thread per context. It's needed by fallback.

Since Velox backend can fully offload lots of queries, it will be interesting to investigate multiple threads per context. We can set task.cores to more than 1 then set to Velox's backend threads. Then use larger partition and see how the performance looks like.

@zhztheplayer Is there any specific consideration on memory management side?

**Comment by zhztheplayer** [comment]
> @zhztheplayer Is there any specific consideration on memory management side?

There should be no hard limitation from Spark's / Gluten's memory management system against using multiple threads in the same Spark task. There were only a few of lock issues that have been resolved already. The only rule is we'd report the allocations to the task's memory manager, and this is something against the approach running a background execution thread pool for all executor's tasks.

> We can set task.cores to more than 1 then set to Velox's backend threads.

Velox's top-level threading model should be designed for Presto's push model. Which doesn't allow Spark's use case that creates an iterator over Velox's task. The parallelism we can easily adopt from Velox is the in-pipeline background thread pools, for example, parallel join build or scan pre-fetch. But again using the thread pools will require us to handle the memory allocation carefully speaking of Spark's memory manager model.

Having said that, there could be a hacky way to combine Spark's pull model with Velox's push model, which requires creating a queue in between them as a data broker. We can do some researches and PoCs to see if this way really brings performance advantages, if it does, we could consider doing refactors on our memory management code and other arch code to switch to it. But it's a big deal and is something more or less violates Spark's design, so we may need to think more before coding.

**Comment by Yohahaha** [comment]
I think we need fix and reenable `spark.gluten.sql.columnar.backend.velox.IOThreads`.

**Comment by zhli1142015** [comment]
> > I think we need fix and reenable spark.gluten.sql.columnar.backend.velox.IOThreads.
> 
> I guess the option can be enabled now since we had solved the related lock problems in code. But I may not be capable to test that in production cases. Helps are appreciated. cc @zhli1142015

I saw the related bug is still open, id this already solved?
Thanks.
https://github.com/apache/incubator-gluten/issues/7161

**Comment by zhztheplayer** [comment]
> I saw the related bug is still open, should we close it also?

Yes, I am closing it. Could reopen once we found it still applies.

**Comment by FelixYBW** [comment]
> I think we need fix and reenable `spark.gluten.sql.columnar.backend.velox.IOThreads`.

we should enable it anyway even in single thread model. It's used to solve IO bottleneck.

**Comment by FelixYBW** [comment]
> Having said that, there could be a hacky way to combine Spark's pull model with Velox's push model, which requires creating a queue in between them as a data broker. We can do some researches and PoCs to see if this way really brings performance advantages, if it does, we could consider doing refactors on our memory management code and other arch code to switch to it. But it's a big deal and is something more or less violates Spark's design, so we may need to think more before coding.

Let's plan it in 2025.

**Comment by surnaik** [comment]
Do we plan to also do the opposite? - multiple task per thread. Currently, we depend on `spark.task.cpus` which is an integer, hence allows either one or more cpu cores per task. For I/O intensive tasks, it would be beneficial to fit more task per CPU.

**Comment by FelixYBW** [comment]
> Do we plan to also do the opposite? - multiple task per thread. Currently, we depend on `spark.task.cpus` which is an integer, hence allows either one or more cpu cores per task. For I/O intensive tasks, it would be beneficial to fit more task per CPU.

It's the over commit of CPU. We need to hack spark to support task.cpus=0.5 etc. Actually we can achieve the similar goal by adding multiple IO threads. you may try something like:

--conf spark.gluten.sql.columnar.backend.velox.IOThreads=24 \
--conf spark.gluten.sql.columnar.backend.velox.prefetchRowGroups=1 \
--conf spark.gluten.sql.columnar.backend.velox.loadQuantum=268435456 \
--conf spark.gluten.sql.columnar.backend.velox.maxCoalescedBytes=67108864 \
--conf spark.gluten.sql.columnar.backend.velox.SplitPreloadPerDriver=0 \
